### ISSUE TITLE
JavadocParagraph now emits errors for comments that are ignored by javadoc

### ISSUE THREAD

**Issue by ewencp** [checkstyle/checkstyle#7535]
Check documentation: https://checkstyle.org/config_javadoc.html#JavadocParagraph

```
/var/tmp $ javac InputJavadocParagraphInvalidLocation.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocParagraph"/>
  </module>
</module>

/var/tmp $ cat InputJavadocParagraphInvalidLocation.java
/**
 * This is a comment in the header of the file.
 * It might be something like a copyright notice.
 *
 * It doesn't conform to Javadoc specs and isn't expected to.
 * Javadoc will ignore it anyway.
 */

package com.puppycrawl.tools.checkstyle.checks.javadoc.javadocparagraph;

/**
 * Some Javadoc.
 *
 * This is incorrectly formatted and should generate an error.
 */
class InputJavadocParagraphInvalidLocation {
}


/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar  checkstyle-8.29-all.jar -c TestConfig.xml \
 src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphInvalidLocation.java
Starting audit...
[ERROR] .../InputJavadocParagraphInvalidLocation.java:3: Empty line should be followed by <p> tag on the next line. [JavadocParagraph]
[ERROR] .../InputJavadocParagraphInvalidLocation.java:12: Empty line should be followed by <p> tag on the next line. [JavadocParagraph]
Audit done.
Checkstyle ends with 2 errors.
```

---------------

Expected (previous) behavior: first error on line 3 not emitted, second error on line 12 is valid.

In earlier versions of checkstyle, some comments that start with javadoc-like `/**` were ignored by certain checks. I believe this changed in https://github.com/checkstyle/checkstyle/issues/6516 in order to catch more invalid javadoc comments, but had the side effect of also causing JavadocParagraphCheck to cover more comments, even if javadoc would ignore them anyway. In a project using JavadocParagraphCheck but not InvalidJavadocPosition, this causes checkstyle to emit new errors. The case I ran into specifically was the use of `/**` to start copyright header comments, where there was no concern about javadoc validity since it wouldn't be processed by javadoc anyway.

Since projects that want the functionality can use InvalidJavadocPosition to catch the location issue, which would then have them either make it non-javadoc or move it to a valid location (at which point JavadocParagraphCheck *should* kick in), I think the check for whether a comment is a javadoc could be refined to be false in these cases. Specifically, for javadoc comments above `package`, they are really only valid in package-info.java files, which shouldn't contain any content other than the javadoc, annotations, and package name.

**Comment by romani** [comment]
@ewencp , I see some commits are detected by GitHub, please do not rush with fix till issue is approved. CI will not accept PR to not approved issue.

This is gray area , we discussed such cases, we have resolution somewhere in issue/PR discussion. I will reply a bit later.

**Comment by romani** [comment]
@ewencp , I see some commits are detected by GitHub, please do not rush with fix till issue is approved. CI will not accept PR to not approved issue.

This is gray area , we discussed such cases, we have resolution somewhere in issue/PR discussion. I will reply a bit later.

**Comment by ewencp** [comment]
Sure thing. I read contribution guidelines and did not file a PR until you mark it approved, but I'll need to work around this even if the patch is not accepted upstream.

**Comment by ewencp** [comment]
Possibly worth some additional context -- I actually have 2 fixes, but one is highly specific to the specific case I encountered (i.e. specific to comments starting on line 1, which for us is always the copyright header), but is completely backwards compatible since it introduces a config for the JavadocParagraphCheck. The one I pushed that you see commits for (and which is already tested against a bunch of the repos in checkstyle-tester) seemed better for upstream since it seems more general and doesn't add any new configs.

**Comment by pbludov** [comment]
Although this is not [documented](https://docs.oracle.com/en/java/javase/11/docs/specs/doc-comment-spec.html), the Javadoc comments on package are ignored in all files except the `package-info.java`.

There is an issue with [InvalidJavadocPosition](https://checkstyle.org/config_javadoc.html#InvalidJavadocPosition). Such comments are invalid and should be reported.

As for `JavadocParagraph` check, there is no bugs - it works as expected. The comment is unformatted and the check reports a violation for it.

*Maybe* some Javadoc checks should ignore misplaced Javadoc comments. It is possible to add a property to control such behaviour.

**Comment by romani** [comment]
>In earlier versions of checkstyle, some comments that start with javadoc-like /** were ignored by certain checks.

Yes, it was like this because all Checks that are created before version 6 were based on parsing by regexp and had tons of bugs. After version 6, we start to do javadoc parsing by grammar. Unfortunately there is not public grammar for javadoc at all. `javadoc` tool is full of bugs too, for many many years user unintentionally used mistake in javadoc content in huge amount of files. So it placed checkstyle in difficult situation to decide what is right and what is not, and always consider ( we not always follow it ) javadoc tool behavior.
javadoc tool was created to generate html files so validation was done only on critical items. Checkstyle goal is opposite - full focus on validation. So checkstyle allow itself be more demanding on items which are ignored by tool.
That is why we are strict on type of multiline commit. If it is `/**` it is javadoc, and it should follow all rules.

It is another point how to ease user experience who do not want to change header and ignore it.
By default we are strict, to let users that do not misuse javadoc comments.
But we should have some way to let user do extra config update and ignore particular cases.

As option you can update your config and place https://checkstyle.org/config_filters.html#SuppressionSingleFilter that will ignore violations from such Check, on specific range of line numbers.


I will extend my reply comment a bit later ....

**Comment by romani** [comment]
workaround, attention to `<property name="lines" value="1-5"/>` it should be changed to match your header size:
```
$ cat Test.java 
/**
 * something like a copyright notice.
 *
 * It doesn't conform to Javadoc specs and isn't expected to.
 * Javadoc will ignore it anyway.
 */
package com.javadocparagraph;


/**
 * Some Javadoc.
 *
 * This is incorrectly formatted and should generate an error.
 */
public class Test {}

$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocParagraph"/>
  </module>
  <module name="SuppressionSingleFilter">
    <property name="checks" value="JavadocParagraph"/>
    <property name="files" value=".*\.java"/>
    <property name="lines" value="1-5"/>
  </module>
</module>

$ java -jar /var/tmp/checkstyle-8.29-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] /var/tmp/Test.java:12: Empty line should be followed by <p> tag on the next line. [JavadocParagraph]
Audit done.
Checkstyle ends with 1 errors.

```

or a bit more flexible (offCommentFormat and onCommentFormat, need to be adjusted to match your header first and last sentence.):
```
$ cat Test.java 
/**
 * something like a copyright notice.
 *
 * It doesn't conform to Javadoc specs and isn't expected to.
 * Javadoc will ignore it anyway.
 */
package com.javadocparagraph;


/**
 * Some Javadoc.
 *
 * This is incorrectly formatted and should generate an error.
 */

$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocParagraph"/>
  </module>
  <module name="SuppressWithPlainTextCommentFilter">
    <property name="offCommentFormat" value="something like a copyright notice."/>
    <property name="onCommentFormat" value="Javadoc will ignore it anyway."/>
    <property name="checkFormat" value="JavadocParagraph"/>
  </module>
</module>

$ java -jar /var/tmp/checkstyle-8.29-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] /var/tmp/Test.java:12: Empty line should be followed by <p> tag on the next line. [JavadocParagraph]
Audit done.
Checkstyle ends with 1 errors.
``

Unfortunately for now Xpath filter does not support javadoc.

**Comment by romani** [comment]
I am closing this issue, suppression should be used.
